storage/innobase/xtrabackup/src/xtrabackup.cc

```diff
@@ -6240,17 +6222,25 @@ redo directory
 @return true if any error occured or false  */
 static bool xtrabackup_close_temp_log(bool clear_flag) {
   char redo_folder[FN_REFLEN];
-  Log_files_context log_files_ctx;
+  const char *backup_dir = !xtrabackup_incremental_dir
+                               ? xtrabackup_target_dir
+                               : xtrabackup_incremental_dir;
 
-  if (!xtrabackup_incremental_dir) {
-    log_files_ctx =
-        Log_files_context{xtrabackup_target_dir, Log_files_ruleset::CURRENT};
-    sprintf(redo_folder, "%s/%s", xtrabackup_target_dir, LOG_DIRECTORY_NAME);
-  } else {
-    log_files_ctx = Log_files_context{xtrabackup_incremental_dir,
-                                      Log_files_ruleset::CURRENT};
-    sprintf(redo_folder, "%s/%s", xtrabackup_incremental_dir,
-            LOG_DIRECTORY_NAME);
+  Log_files_context log_files_ctx =
+      Log_files_context{backup_dir, Log_files_ruleset::CURRENT};
+  sprintf(redo_folder, "%s/%s", backup_dir, LOG_DIRECTORY_NAME);
+
+  bool valid = xb_log_files_validate_creators(backup_dir);
+
+  if (!valid) {
+    xb::error()
+        << "Unable to validate the redo log files at the end of prepare";
+    return (true);
+  }
+
+  if (xb_generated_redo && !clear_flag) {
+    xb::info() << "prepare generated new redo log files. Skipping rename. ";
+    return (false);
   }
 
   if (xtrabackup_logfile_is_renamed) {
```

Comment:
Hmm, so it looks like we keep xtrbackup_logfile in incremental backup dirs and also in case of full backups that didn't generated new redo files.

When the last prepare is done and new redo files are generated, we should avoid rename back. This should be fixed.